validator_client/src/duties_service.rs

```diff
@@ -1,758 +1,755 @@
+    // defensive regardless.
+    drop(attesters_map);
 
-    /// Attempt to download the duties of all managed validators for the given `request_epoch`. The
-    /// `current_epoch` should be a local reading of the slot clock.
-    async fn update_epoch(
-        self,
-        current_epoch: Epoch,
-        request_epoch: Epoch,
-        spec: &ChainSpec,
-    ) -> Result<(), String> {
-        let log = self.context.log();
-
-        let maybe_require_synced = if self.allow_unsynced_beacon_node {
-            RequireSynced::No
-        } else {
-            RequireSynced::Yes
-        };
-
-        let mut new_validator = 0;
-        let mut new_epoch = 0;
-        let mut new_proposal_slots = 0;
-        let mut identical = 0;
-        let mut replaced = 0;
-        let mut invalid = 0;
-
-        // Determine which pubkeys we already know the index of by checking the duties store for
-        // the current epoch.
-        let pubkeys: Vec<(PublicKey, Option<u64>)> = self
-            .validator_store
-            .voting_pubkeys()
-            .into_iter()
-            .map(|pubkey| {
-                let index = self.store.get_index(&pubkey, current_epoch);
-                (pubkey, index)
-            })
-            .collect();
+    Ok(())
+}
 
-        let mut validator_subscriptions = vec![];
-        let pubkeys_ref = &pubkeys;
-        let remote_duties: Vec<ValidatorDuty> = match self
+/// Download the proposer duties for the current epoch and store them in `duties_service.proposers`.
+/// If there are any proposer for this slot, send out a notification to the block proposers.
+///
+/// ## Note
+///
+/// This function will potentially send *two* notifications to the `BlockService`; it will send a
+/// notification initially, then it will download the latest duties and send a *second* notification
+/// if those duties have changed. This behaviour simultaneously achieves the following:
+///
+/// 1. Block production can happen immediately and does not have to wait for the proposer duties to
+///    download.
+/// 2. We won't miss a block if the duties for the current slot happen to change with this poll.
+///
+/// This sounds great, but is it safe? Firstly, the additional notification will only contain block
+/// producers that were not included in the first notification. This should be safety enough.
+/// However, we also have the slashing protection as a second line of defence. These two factors
+/// provide an acceptable level of safety.
+///
+/// It's important to note that since there is a 0-epoch look-ahead (i.e., no look-ahead) for block
+/// proposers then it's very likely that a proposal for the first slot of the epoch will need go
+/// through the slow path every time. I.e., the proposal will only happen after we've been able to
+/// download and process the duties from the BN. This means it is very important to ensure this
+/// function is as fast as possible.
```

Comment:
I agree with everything written here, but I wonder if at some point in the future it might be worth experimenting with polling partway through the end of the previous slot (i.e. once the dependent-root block has arrived). Or is this obsoleted by some future API on the BN that pushes re-org info to the VC?

Reply:
Yeah, I totally agree that's worth experimenting with in the future. I don't know of a standard API event that would work well here, since we make no guarantees about the proposer shuffling being known for the next epoch after we get a block. I think the best we could do is add our own event (not super keen on this, tbh) or just make an assumption about when the BN should already know then shuffling and then poll then.

One of the benefits of checking the proposer shuffling early would be that we prime the caches for the scenario where the last block of the epoch is missed.

Reply:
Hey, wondering if there has been further discussions on this. Similar logic is also implemented in Lodestar but facing some issues due to that https://github.com/ChainSafe/lodestar/issues/5314. Skipping a slot seems to cause more harm then good in a lot of cases. Lodestar applies this logic also for other routines such as attestations and sync committe